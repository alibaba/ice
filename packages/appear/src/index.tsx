import { Children, useRef, useEffect, useCallback } from 'react';
import type { Ref } from 'react';
import { isFunction } from './type';
import { observerElement, VisibilityChangeEvent } from './visibility';

function VisibilityChange(props: any) {
  const {
    onAppear,
    onDisappear,
    children,
  } = props;

  const defaultRef: Ref<Node> = useRef<Node>();
  const ref: Ref<Node> = (children && children.ref) ? children.ref : defaultRef;

  const listen = useCallback((eventName: string, handler: Function) => {
    const { current } = ref;
    // Rax components will set custom ref by useImperativeHandle.
    // So We should get eventTarget by _nativeNode.
    // https://github.com/raxjs/rax-components/blob/master/packages/rax-textinput/src/index.tsx#L151
    if (current && isFunction(handler)) {
      const eventTarget = current._nativeNode || current;
      observerElement(eventTarget as Element);
      eventTarget.addEventListener(eventName, handler);
    }
    return () => {
      const { current } = ref;
      if (current) {
        const eventTarget = current._nativeNode || current;
        eventTarget.removeEventListener(eventName, handler);
      }
    };
  }, [ref]);

  useEffect(() => listen(VisibilityChangeEvent.appear, onAppear), [ref, onAppear, listen]);
  useEffect(() => listen(VisibilityChangeEvent.disappear, onDisappear), [ref, onDisappear, listen]);

  return Children.only({ ...children, ref });
}

export default VisibilityChange;
