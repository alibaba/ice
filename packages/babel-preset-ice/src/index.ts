interface IOpts {
  typescript?: boolean;
  env?: object;
  react?: boolean | object;
}
/**
 * babel config
 */
function resolvePlugin(plugins) {
  return plugins.filter(Boolean).map((plugin) => {
    if (Array.isArray(plugin)) {
      const [pluginName, ...args] = plugin;
      return [require.resolve(pluginName), ...args];
    }
    return require.resolve(plugin);
  });
}

export default (opts: IOpts = {}) => {
  const plugins = [
    // Stage 0
    '@babel/plugin-proposal-function-bind',
    // Stage 1
    '@babel/plugin-proposal-export-default-from',
    '@babel/plugin-proposal-logical-assignment-operators',
    ['@babel/plugin-proposal-optional-chaining', { loose: false }],
    ['@babel/plugin-proposal-pipeline-operator', { proposal: 'minimal' }],
    ['@babel/plugin-proposal-nullish-coalescing-operator', { loose: false }],
    '@babel/plugin-proposal-do-expressions',
    // Stage 2
    ['@babel/plugin-proposal-decorators', { legacy: true }],
    '@babel/plugin-proposal-function-sent',
    '@babel/plugin-proposal-export-namespace-from',
    '@babel/plugin-proposal-numeric-separator',
    '@babel/plugin-proposal-throw-expressions',
    // Stage 3
    '@babel/plugin-syntax-dynamic-import',
    '@babel/plugin-syntax-import-meta',
    '@babel/plugin-proposal-json-strings',
  ];

  return {
    presets: resolvePlugin([
      opts.env && [
        '@babel/preset-env', opts.env,
      ],
      opts.typescript && '@babel/preset-typescript',
      opts.react && (typeof opts.react === 'boolean'
        ? '@babel/preset-react'
        : ['@babel/preset-react', opts.react]
      ),
    ]),
    plugins: resolvePlugin(plugins),
  };
};
