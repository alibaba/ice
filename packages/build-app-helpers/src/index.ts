export { default as getMpaEntries } from './getMpaEntries';
export { default as getRoutesByAppJson } from './getRoutesByAppJson';
export { default as getEntriesByRoute } from './getEntriesByRoute';
export { default as formatPath } from './formatPath';
export { default as validation } from './validation';
export { default as unionBy } from './unionBy';
export { default as injectTransformRuntime } from './injectTransformRuntime';
export { default as checkExportDefaultDeclarationExists } from './checkExportDefaultDeclarationExists';
export { default as getSassImplementation } from './getSassImplementation';
export { default as getLessImplementation } from './getLessImplementation';
export { default as redirectImport } from './redirectImport';
