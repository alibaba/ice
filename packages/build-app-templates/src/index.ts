import * as path from 'path';

type FrameworkType = 'rax' | 'react';

export function getTemplate(templateName: string, framework?: FrameworkType) {
  if (!framework) return path.join(__dirname, 'templates/common', `${templateName}.ejs`);
  return path.join(__dirname, 'templates', framework, `${templateName}.ejs`);
}

export function getCommonTemplateDir() {
  return path.join(__dirname, 'templates/common');
}

export function getFrameworkTemplateDir(framework: FrameworkType) {
  return path.join(__dirname, `templates/${framework}`);
}
