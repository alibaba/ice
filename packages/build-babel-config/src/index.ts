import getBabelPreset from './preset-ice';

export { getBabelPreset };
export default () => {
  return getBabelPreset({
    react: {},
    typescript: true,
    env: {
      modules: false,
      useBuiltIns: 'entry',
      // Set latest polyfills (until ECMAScript 2021). https://unpkg.alibaba-inc.com/browse/core-js-compat@3.18.3/modules-by-versions.json
      corejs: '3.7',
    }
  });
};
