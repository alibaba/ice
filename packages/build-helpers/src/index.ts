import getEntries from './getEntries';
import getRoutes from './getRoutes';

export {
  getEntries,
  getRoutes
};
