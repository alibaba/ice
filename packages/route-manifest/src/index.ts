// based on https://github.com/remix-run/remix/blob/main/packages/remix-dev/config/routesConvention.ts
import fs from 'fs';
import path from 'path';
import minimatch from 'minimatch';
import {
  createComponentName,
  removeLastLayoutStrFromId,
  createRouteId,
  defineRoutes,
  normalizeSlashes,
  createFileId,
} from './routes.js';
import type {
  RouteManifest,
  DefineRouteFunction,
  NestedRouteManifest,
  ConfigRoute,
  DefineRoutesOptions,
  DefineExtraRoutes,
} from './routes.js';

export type {
  RouteManifest,
  NestedRouteManifest,
  DefineRouteFunction,
  ConfigRoute,
  DefineExtraRoutes,
  DefineRoutesOptions,
};

export interface RouteItem {
  path: string;
  component: string;
  children?: RouteItem[];
}

const validRouteChar = ['-', '\\w', '/', ':', '*', '\\[', '\\]', '\\.'];

const routeModuleExts = [
  '.js',
  '.jsx',
  '.ts',
  '.tsx',
  // TODO: 暂不支持 .md/.mdx 文件，需要工程配合
  // '.md',
  // '.mdx',
];

export function isRouteModuleFile(filename: string): boolean {
  return routeModuleExts.includes(path.extname(filename));
}


export function generateRouteManifest(
  rootDir: string,
  ignoreFiles: string[] = [],
  defineExtraRoutesQueue?: DefineExtraRoutes[],
  routeConfig?: RouteItem[],
) {
  const srcDir = path.join(rootDir, 'src');
  const routeManifest: RouteManifest = {};
  // 2. find routes in `src/pages` directory
  if (fs.existsSync(path.resolve(srcDir, 'pages'))) {
    const conventionalRoutes = defineConventionalRoutes(
      rootDir,
      {
        routeManifest,
        nestedRouteManifest: formatNestedRouteManifest(routeManifest),
      },
      ignoreFiles,
    );

    for (const key of Object.keys(conventionalRoutes)) {
      const route = conventionalRoutes[key];
      routeManifest[route.id] = {
        ...route,
        parentId: route.parentId || undefined,
      };
    }
  }
  // 3. add extra routes from user config
  if (Array.isArray(defineExtraRoutesQueue)) {
    defineExtraRoutesQueue.forEach((defineExtraRoutes) => {
      if (defineExtraRoutes) {
        const extraRoutes = defineRoutes(
          defineExtraRoutes,
          {
            routeManifest,
            nestedRouteManifest: formatNestedRouteManifest(routeManifest),
          },
        );
        for (const key of Object.keys(extraRoutes)) {
          const route = extraRoutes[key];
          routeManifest[route.id] = {
            ...route,
            parentId: route.parentId || undefined,
          };
        }
      }
    });
  }

  // Add routes by routes config.
  if (routeConfig) {
    routeConfig.forEach((routeItem) => {
      const routes = parseRoute(routeItem);
      routes.forEach((route) => {
        routeManifest[route.id] = route;
      });
    });
  }

  return routeManifest;
}

export function parseRoute(routeItem: RouteItem, parentId?: string, parentPath?: string) {
  const routes = [];
  const { path: routePath, component, children } = routeItem;
  const id = createRouteId(component, routePath, parentPath);
  let index;
  const currentPath = path.join(parentPath || '/', routePath).split(path.sep).join('/');
  const isRootPath = currentPath === '/';
  if (!children && isRootPath) {
    index = true;
  }
  const route: ConfigRoute = {
    // An absolute child route path must start with the combined path of all its parent routes
    // Replace the first slash with an empty string to compatible with the route definintion, e.g. /foo
    path: parentId && routePath !== '/' ? routePath.replace(/^\//, '') : routePath,
    index,
    id,
    parentId,
    file: component,
    componentName: createComponentName(component),
    layout: !!children,
  };
  routes.push(route);
  if (children) {
    children.forEach((childRoute) => {
      routes.push(...parseRoute(childRoute, id, currentPath));
    });
  }
  return routes;
}

export function formatNestedRouteManifest(routeManifest: RouteManifest, parentId?: string): NestedRouteManifest[] {
  return Object.keys(routeManifest)
    .filter(key => routeManifest[key].parentId === parentId)
    .map(key => {
      const route = {
        ...routeManifest[key],
      } as NestedRouteManifest;
      const children = formatNestedRouteManifest(routeManifest, route.id);
      if (children.length > 0) route.children = children;
      return route;
    });
}

function defineConventionalRoutes(
  rootDir: string,
  options: DefineRoutesOptions,
  ignoredFilePatterns?: string[],
): RouteManifest {
  const files: { [routeId: string]: string } = {};
  // 1. find all route components in src/pages
  visitFiles(
    path.join(rootDir, 'src', 'pages'),
    file => {
      if (
        ignoredFilePatterns &&
        ignoredFilePatterns.some(pattern => minimatch(file, pattern))
      ) {
        return;
      }

      if (isRouteModuleFile(file)) {
        let fileId = createFileId(file);
        files[fileId] = file;
        return;
      }
    },
  );

  const fileIds = Object.keys(files).sort(byLongestFirst);

  const uniqueRoutes = new Map<string, string>();

  // 2. recurse through all routes using the public defineRoutes() API
  function defineNestedRoutes(
    defineRoute: DefineRouteFunction,
    options: DefineRoutesOptions,
    parentId?: string,
  ): void {
    const childFileIds = fileIds.filter((id) => {
      const parentFileId = findParentFileId(fileIds, id);
      return parentFileId === parentId;
    });

    for (let fileId of childFileIds) {
      const parentRoutePath = removeLastLayoutStrFromId(parentId) || '';
      const routePath: string | undefined = createRoutePath(
        // parentRoutePath = 'home', routeId = 'home/me', the new routeId is 'me'
        // in order to escape the child route path is absolute path
        fileId.slice(parentRoutePath.length + (parentRoutePath ? 1 : 0)),
      );
      const routeFilePath = normalizeSlashes(path.join('src', 'pages', files[fileId]));
      if (RegExp(`[^${validRouteChar.join('')}]+`).test(routePath)) {
        throw new Error(`invalid character in '${routeFilePath}'. Only support char: ${validRouteChar.join(', ')}`);
      }
      const isIndexRoute = fileId === 'index' || fileId.endsWith('/index');
      const fullPath = createRoutePath(fileId);
      const uniqueRouteId = (fullPath || '') + (isIndexRoute ? '?index' : '');

      if (uniqueRouteId) {
        if (uniqueRoutes.has(uniqueRouteId)) {
          throw new Error(
            `Path ${JSON.stringify(fullPath)} defined by route ${JSON.stringify(routeFilePath)}
            conflicts with route ${JSON.stringify(uniqueRoutes.get(uniqueRouteId))}`,
          );
        } else {
          uniqueRoutes.set(uniqueRouteId, fileId);
        }
      }

      if (isIndexRoute) {
        let invalidChildRoutes = fileIds.filter(
          (id) => findParentFileId(fileIds, id) === fileId,
        );

        if (invalidChildRoutes.length > 0) {
          throw new Error(
            `Child routes are not allowed in index routes. Please remove child routes of ${fileId}`,
          );
        }

        defineRoute(routePath, files[fileId], {
          index: true,
        });
      } else {
        defineRoute(routePath, files[fileId], () => {
          defineNestedRoutes(defineRoute, options, fileId);
        });
      }
    }
  }

  return defineRoutes(defineNestedRoutes, options);
}

const escapeStart = '[';
const escapeEnd = ']';

export function createRoutePath(routeId: string): string | undefined {
  let result = '';
  let rawSegmentBuffer = '';

  let inEscapeSequence = 0;
  let skipSegment = false;

  const partialRouteId = removeLastLayoutStrFromId(routeId);

  for (let i = 0; i < partialRouteId.length; i++) {
    const char = partialRouteId.charAt(i);
    const lastChar = i > 0 ? partialRouteId.charAt(i - 1) : undefined;
    const nextChar = i < partialRouteId.length - 1 ? partialRouteId.charAt(i + 1) : undefined;

    const isNewEscapeSequence = !inEscapeSequence && char === escapeStart && lastChar !== escapeStart;
    const isCloseEscapeSequence = inEscapeSequence && char === escapeEnd && nextChar !== escapeEnd;

    if (skipSegment) {
      if (char === '/' || char === '.' || char === path.win32.sep) {
        skipSegment = false;
      }
      continue;
    }
    // We try to get the string in the escape sequence. For example, try to get the `index` string in `[index]`.
    if (isNewEscapeSequence) {
      inEscapeSequence++;
      continue;
    }

    if (isCloseEscapeSequence) {
      inEscapeSequence--;
      continue;
    }

    if (inEscapeSequence) {
      result += char;
      continue;
    }

    if (char === '/' || char === path.win32.sep || char === '.') {
      if (rawSegmentBuffer === 'index' && result.endsWith('index')) {
        result = result.replace(/\/?index$/, '');
      } else {
        result += '/';
      }
      rawSegmentBuffer = '';
      continue;
    }

    rawSegmentBuffer += char;

    if (char === '$') {
      result += typeof nextChar === 'undefined' ? '*' : ':';
      continue;
    }

    result += char;
  }

  if (rawSegmentBuffer === 'index' && result.endsWith('index')) {
    result = result.replace(/\/?index$/, '');
  }

  return result || undefined;
}

function findParentFileId(
  routeIds: string[],
  childRouteId: string,
): string | undefined {
  return routeIds.find((id) => {
    // childRouteId is `pages/about` and id is `pages/layout` will match
    // childRouteId is `pages/about/index` and id is `pages/about/layout` will match
    return childRouteId !== id && id.endsWith('layout') && childRouteId.startsWith(`${id.slice(0, id.length - 'layout'.length)}`);
  });
}

function byLongestFirst(a: string, b: string): number {
  return b.length - a.length;
}

function visitFiles(
  dir: string,
  visitor: (file: string) => void,
  baseDir = dir,
): void {
  for (let filename of fs.readdirSync(dir)) {
    let file = path.resolve(dir, filename);
    let stat = fs.lstatSync(file);

    if (stat.isDirectory()) {
      visitFiles(file, visitor, baseDir);
    } else if (stat.isFile()) {
      visitor(normalizeSlashes(path.relative(baseDir, file)));
    }
  }
}
