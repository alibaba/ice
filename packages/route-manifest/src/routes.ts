// based on https://github.com/remix-run/remix/blob/main/packages/remix-dev/config/routes.ts

import { win32, join } from 'path';

export interface ConfigRoute {
  /**
   * The unique id for this route, named like its `file` but without the
   * extension. So `src/pages/gists/$username.jsx` will have an `id` of
   * `routes/gists/$username`.
   */
  id: string;
  /**
   * The path to the entry point for this route, relative to src/pages
   */
  file: string;
  /**
   * current route's component. `src/pages/gist.jsx` will have an `componentName` of `PageGist`
   */
  componentName: string;
  /**
   * The unique `id` for this route's parent route, if there is one.
   */
  parentId?: string;
  /**
   * The path this route uses to match on the URL pathname.
   */
  path: string;
  /**
   * Should be `true` if it is an index route. This disallows child routes.
   */
  index?: boolean;
  /**
   * Should be `true` if route is layout component.
   */
  layout?: boolean;
  /**
   * module exports key of route
   */
  exports?: string[];
}

export interface DefineRouteOptions {
  /**
   * Should be `true` if this is an index route that does not allow child routes.
   */
  index?: boolean;
}

interface DefineRouteChildren {
  (): void;
}

export interface DefineRouteFunction {
  (
    /**
     * The path this route uses to match the URL pathname.
     */
    path: string | undefined,

    /**
     * The path to the file that exports the React component rendered by this
     * route as its default export, relative to the `app` directory.
     */
    file: string,

    /**
     * Options for defining routes, or a function for defining child routes.
     */
    optionsOrChildren?: DefineRouteOptions | DefineRouteChildren,

    /**
     * A function for defining child routes.
     */
    children?: DefineRouteChildren,
  ): void;
}
export interface RouteManifest {
  [routeId: string]: ConfigRoute;
}

export interface NestedRouteManifest extends ConfigRoute {
  children?: ConfigRoute[];
}

export interface DefineRoutesOptions {
  routeManifest: RouteManifest;
  nestedRouteManifest: NestedRouteManifest[];
}

export type DefineExtraRoutes = (
  defineRoute: DefineRouteFunction,
  options: DefineRoutesOptions,
) => void;

export function defineRoutes(
  callback: (defineRoute: DefineRouteFunction, options: DefineRoutesOptions) => void,
  options: DefineRoutesOptions,
) {
  const routes: RouteManifest = Object.create(null);
  const parentRoutes: ConfigRoute[] = [];
  let alreadyReturned = false;

  const defineRoute: DefineRouteFunction = (path, file, optionsOrChildren, children) => {
    if (alreadyReturned) {
      throw new Error(
        'You tried to define routes asynchronously but started defining ' +
          'routes before the async work was done. Please await all async ' +
          'data before calling `defineRoutes()`',
      );
    }

    let options: DefineRouteOptions;
    if (typeof optionsOrChildren === 'function') {
      // route(path, file, children)
      options = {};
      children = optionsOrChildren;
    } else {
      // route(path, file, options, children)
      // route(path, file, options)
      options = optionsOrChildren || {};
    }
    const parentRoute = parentRoutes.length > 0
      ? parentRoutes[parentRoutes.length - 1]
      : undefined;

    const id = createRouteId(file, path, parentRoute?.path, options.index);

    const route: ConfigRoute = {
      path,
      index: options.index ? true : undefined,
      id,
      parentId: parentRoute ? parentRoute.id : undefined,
      file,
      componentName: createComponentName(file),
      layout: id.endsWith('layout'),
    };

    routes[route.id] = route;

    if (children) {
      parentRoutes.push(route);
      children();
      parentRoutes.pop();
    }
  };

  callback(defineRoute, options);

  alreadyReturned = true;

  return routes;
}

export function createRouteId(
  file: string,
  path?: string,
  parentPath?: string,
  index?: boolean,
) {
  return normalizeSlashes(join(
    parentPath || '',
    path || (index ? '/' : ''),
    stripFileExtension(file).endsWith('layout') ? 'layout' : '',
  ));
}

export function createFileId(file: string) {
  return normalizeSlashes(stripFileExtension(file));
}

export function normalizeSlashes(file: string) {
  return file.split(win32.sep).join('/');
}

function stripFileExtension(file: string) {
  return file.replace(/\.[a-z0-9]+$/i, '');
}

export function createComponentName(file: string) {
  return createFileId(file).split('/')
    .map((item: string) => item.toLowerCase())
    .join('-');
}

/**
 * remove `/layout` str if the routeId has it
 *
 * 'layout' -> ''
 * 'About/layout' -> 'About'
 * 'About/layout/index' -> 'About/layout/index'
 */
export function removeLastLayoutStrFromId(id?: string) {
  const layoutStrs = ['/layout', 'layout'];
  const currentLayoutStr = layoutStrs.find(layoutStr => id?.endsWith(layoutStr));
  return currentLayoutStr ? id.slice(0, id.length - currentLayoutStr.length) : id;
}
