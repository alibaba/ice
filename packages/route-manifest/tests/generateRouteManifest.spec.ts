import path from 'path';
import { fileURLToPath } from 'url';
import { expect, test, describe } from 'vitest';
import { generateRouteManifest } from '../src/index';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const fixturesDir = path.join(__dirname, 'fixtures');

describe('generateRouteManifest function', () => {
  test('basic-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'basic-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('dynamic-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'dynamic-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('layout-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'layout-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('doc-delimeters-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'dot-delimeters-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('splat-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'splat-routes'));
    expect(routeManifest).toMatchSnapshot();
  });

  test('invalid-routes', () => {
    expect(() => generateRouteManifest(path.join(fixturesDir, 'invalid-routes'))).toThrow('invalid character in \'src/pages/#a.tsx\'. Only support char: -, \\w, /');
  });

  test('ignore-routes', () => {
    const routeManifest = generateRouteManifest(path.join(fixturesDir, 'basic-routes'), ['About/index.tsx']);
    expect(routeManifest).toMatchSnapshot();
  });

  test('define-extra-routes', () => {
    const routeManifest = generateRouteManifest(
      path.join(fixturesDir, 'basic-routes'),
      ['About/index.tsx'],
      (defineRoute) => {
        defineRoute('/about-me', 'About/index.tsx');
      },
    );
    expect(routeManifest).toMatchSnapshot();
  });

  test('escape-routes', () => {
    const routeManifest = generateRouteManifest(
      path.join(fixturesDir, 'escape-routes'),
    );
    expect(routeManifest).toMatchSnapshot();
  });

  test('nested-routes', () => {
    const routeManifest = generateRouteManifest(
      path.join(fixturesDir, 'nested-routes'),
    );
    expect(routeManifest).toMatchSnapshot();
  });
});
