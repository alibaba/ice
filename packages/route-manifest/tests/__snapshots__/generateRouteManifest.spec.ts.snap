// Vitest Snapshot v1

exports[`generateRouteManifest function > basic-routes 1`] = `
{
  "/": {
    "componentName": "index",
    "file": "index.tsx",
    "id": "/",
    "index": true,
    "layout": false,
    "parentId": "layout",
    "path": undefined,
  },
  "About": {
    "componentName": "about-index",
    "file": "About/index.tsx",
    "id": "About",
    "index": true,
    "layout": false,
    "parentId": "layout",
    "path": "About",
  },
  "About/me": {
    "componentName": "about-me-index",
    "file": "About/me/index.tsx",
    "id": "About/me",
    "index": true,
    "layout": false,
    "parentId": "layout",
    "path": "About/me",
  },
  "home": {
    "componentName": "home",
    "file": "home.tsx",
    "id": "home",
    "index": undefined,
    "layout": false,
    "parentId": "layout",
    "path": "home",
  },
  "layout": {
    "componentName": "layout",
    "file": "layout.tsx",
    "id": "layout",
    "index": undefined,
    "layout": true,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function > define-extra-routes 1`] = `
{
  "/": {
    "componentName": "index",
    "file": "index.tsx",
    "id": "/",
    "index": true,
    "layout": false,
    "parentId": "layout",
    "path": undefined,
  },
  "/about-me": {
    "componentName": "about-index",
    "file": "About/index.tsx",
    "id": "/about-me",
    "index": undefined,
    "layout": false,
    "parentId": undefined,
    "path": "/about-me",
  },
  "About/me": {
    "componentName": "about-me-index",
    "file": "About/me/index.tsx",
    "id": "About/me",
    "index": true,
    "layout": false,
    "parentId": "layout",
    "path": "About/me",
  },
  "home": {
    "componentName": "home",
    "file": "home.tsx",
    "id": "home",
    "index": undefined,
    "layout": false,
    "parentId": "layout",
    "path": "home",
  },
  "layout": {
    "componentName": "layout",
    "file": "layout.tsx",
    "id": "layout",
    "index": undefined,
    "layout": true,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function > doc-delimeters-routes 1`] = `
{
  "home/news": {
    "componentName": "home.news",
    "file": "home.news.tsx",
    "id": "home/news",
    "index": undefined,
    "layout": false,
    "parentId": "layout",
    "path": "home/news",
  },
  "layout": {
    "componentName": "layout",
    "file": "layout.tsx",
    "id": "layout",
    "index": undefined,
    "layout": true,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function > dynamic-routes 1`] = `
{
  "/": {
    "componentName": "index",
    "file": "index.tsx",
    "id": "/",
    "index": true,
    "layout": false,
    "parentId": undefined,
    "path": undefined,
  },
  "about": {
    "componentName": "about",
    "file": "about.tsx",
    "id": "about",
    "index": undefined,
    "layout": false,
    "parentId": undefined,
    "path": "about",
  },
  "blog": {
    "componentName": "blog-index",
    "file": "blog/index.tsx",
    "id": "blog",
    "index": true,
    "layout": false,
    "parentId": undefined,
    "path": "blog",
  },
  "blog/:id": {
    "componentName": "blog-$id",
    "file": "blog/$id.tsx",
    "id": "blog/:id",
    "index": undefined,
    "layout": false,
    "parentId": undefined,
    "path": "blog/:id",
  },
}
`;

exports[`generateRouteManifest function > escape-routes 1`] = `
{
  "1.pdf": {
    "componentName": "1[.pdf]",
    "file": "1[.pdf].tsx",
    "id": "1.pdf",
    "index": undefined,
    "layout": false,
    "parentId": undefined,
    "path": "1.pdf",
  },
  "index": {
    "componentName": "[index]",
    "file": "[index].tsx",
    "id": "index",
    "index": undefined,
    "layout": false,
    "parentId": undefined,
    "path": "index",
  },
}
`;

exports[`generateRouteManifest function > ignore-routes 1`] = `
{
  "/": {
    "componentName": "index",
    "file": "index.tsx",
    "id": "/",
    "index": true,
    "layout": false,
    "parentId": "layout",
    "path": undefined,
  },
  "About/me": {
    "componentName": "about-me-index",
    "file": "About/me/index.tsx",
    "id": "About/me",
    "index": true,
    "layout": false,
    "parentId": "layout",
    "path": "About/me",
  },
  "home": {
    "componentName": "home",
    "file": "home.tsx",
    "id": "home",
    "index": undefined,
    "layout": false,
    "parentId": "layout",
    "path": "home",
  },
  "layout": {
    "componentName": "layout",
    "file": "layout.tsx",
    "id": "layout",
    "index": undefined,
    "layout": true,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function > layout-routes 1`] = `
{
  "/": {
    "componentName": "index",
    "file": "index.tsx",
    "id": "/",
    "index": true,
    "layout": false,
    "parentId": "layout",
    "path": undefined,
  },
  "about": {
    "componentName": "about",
    "file": "about.tsx",
    "id": "about",
    "index": undefined,
    "layout": false,
    "parentId": "layout",
    "path": "about",
  },
  "blog/": {
    "componentName": "blog-index",
    "file": "blog/index.tsx",
    "id": "blog/",
    "index": true,
    "layout": false,
    "parentId": "blog/layout",
    "path": undefined,
  },
  "blog/:id": {
    "componentName": "blog-$id",
    "file": "blog/$id.tsx",
    "id": "blog/:id",
    "index": undefined,
    "layout": false,
    "parentId": "blog/layout",
    "path": ":id",
  },
  "blog/layout": {
    "componentName": "blog-layout",
    "file": "blog/layout.tsx",
    "id": "blog/layout",
    "index": undefined,
    "layout": true,
    "parentId": "layout",
    "path": "blog",
  },
  "home/": {
    "componentName": "home-index",
    "file": "home/index.tsx",
    "id": "home/",
    "index": true,
    "layout": false,
    "parentId": "home/layout",
    "path": undefined,
  },
  "home/layout": {
    "componentName": "home-layout-index",
    "file": "home/layout/index.tsx",
    "id": "home/layout",
    "index": true,
    "layout": true,
    "parentId": "home/layout",
    "path": "layout",
  },
  "layout": {
    "componentName": "layout",
    "file": "layout.tsx",
    "id": "layout",
    "index": undefined,
    "layout": true,
    "parentId": undefined,
    "path": undefined,
  },
}
`;

exports[`generateRouteManifest function > nested-routes 1`] = `
{
  "a/b/c": {
    "componentName": "a-b-c",
    "file": "a/b/c.tsx",
    "id": "a/b/c",
    "index": undefined,
    "layout": false,
    "parentId": undefined,
    "path": "a/b/c",
  },
  "d/e/f": {
    "componentName": "d.e.f",
    "file": "d.e.f.tsx",
    "id": "d/e/f",
    "index": undefined,
    "layout": false,
    "parentId": undefined,
    "path": "d/e/f",
  },
}
`;

exports[`generateRouteManifest function > splat-routes 1`] = `
{
  "*": {
    "componentName": "$",
    "file": "$.tsx",
    "id": "*",
    "index": undefined,
    "layout": false,
    "parentId": "layout",
    "path": "*",
  },
  "home": {
    "componentName": "home",
    "file": "home.tsx",
    "id": "home",
    "index": undefined,
    "layout": false,
    "parentId": "layout",
    "path": "home",
  },
  "layout": {
    "componentName": "layout",
    "file": "layout.tsx",
    "id": "layout",
    "index": undefined,
    "layout": true,
    "parentId": undefined,
    "path": undefined,
  },
}
`;
